src/app/Recordings/ArchiveUploadModal.tsx

```diff
@@ -53,44 +55,61 @@ export const ArchiveUploadModal: React.FunctionComponent<ArchiveUploadModalProps
   const [filename, setFilename] = React.useState('' as string | undefined);
   const [uploading, setUploading] = React.useState(false);
   const [rejected, setRejected] = React.useState(false);
+  const [showCancelPrompt, setShowCancelPrompt] = React.useState(false);
+  const [abort, setAbort] = React.useState(new AbortController());
 
-  const reset = () => {
+  const reset = React.useCallback(() => {
     setUploadFile(undefined);
     setFilename('');
     setUploading(false);
     setRejected(true);
-  };
+    setShowCancelPrompt(false);
+    setAbort(new AbortController());
+  }, [setUploadFile, setFilename, setUploading, setRejected, setShowCancelPrompt, setAbort]);
 
-  const handleFileChange = (file, filename) => {
+  const handleFileChange = React.useCallback((file, filename) => {
     setRejected(false);
     setUploadFile(file);
     setFilename(filename);
-  };
+    setShowCancelPrompt(false);
+  }, [setRejected, setUploadFile, setFilename, setShowCancelPrompt]);
 
-  const handleReject = () => {
+  const handleReject = React.useCallback(() => {
     setRejected(true);
-  };
+    setShowCancelPrompt(false);
+  }, [setRejected, setShowCancelPrompt]);
 
-  const handleClose = () => {
-    reset();
-    props.onClose();
-  };
+  const handleClose = React.useCallback(() => {
+    if (uploading) {
+      setShowCancelPrompt(true);
+    } else {
+      reset();
+      props.onClose();
+    }
+  }, [uploading, setShowCancelPrompt, reset]);
 
-  const handleSubmit = () => {
+  const handleSubmit = React.useCallback(() => {
     if (!uploadFile) {
       notifications.warning('Attempted to submit JFR upload without a file selected');
       return;
     }
     setUploading(true);
-    context.api.uploadRecording(uploadFile)
-      .pipe(
-        first(),
-        tap(() => setUploading(false)),
-      )
+    context.api.uploadRecording(uploadFile, abort.signal)
+      .pipe(first())
       .subscribe(handleClose, reset);
-  };
+  }, [context.api, notifications, setUploading, uploadFile, abort, handleClose, reset]);
+
+  const handleAbort = React.useCallback(() => {
+    abort.abort();
+    reset();
+    props.onClose();
+  }, [abort, reset, handleClose]);
 
-  return (
+  return (<>
+    <Prompt
+      when={uploading}
+      message="Are you sure you wish to cancel the file upload?"
```

Comment:
How can I view this prompt message? If I refresh the page while uploading a recording, I see the pop-up dialog "Reload site? Changes you made may not be saved"

Reply:
Refreshing the page would trigger the `onbeforeunload` handler that I explained to Hareet in a comment above, so that message is actually implemented and displayed by the browser itself (hence why it lprobably ooks native and not like a Patternfly component). This `Prompt` component is from the `react-router-dom` module, so it's Router-based - you would need to try to navigate in-app to make it appear. With the Cancel modal in place there isn't really a way to do that though. I can remove this, since I put this in first before the `onbeforeunload` hook, and this is now probably dead code in all actual scenarios, but it seemed like a pretty cheap piece of "insurance" to include in case maybe some browser doesn't display the expected prompt with the `onbeforeunload` hook.

Reply:
That said, if you want to simulate what you might see if such a browser existed that didn't display the expected `onbeforeunload` prompt, you could just comment out the line that assigns a function to that hook in the Api.service. Then after starting a file upload, try to navigate elsewhere in the app, say to the Events view.

Reply:
Makes sense, thanks